test/FilterTest.php

```diff
@@ -3,109 +3,99 @@
+            $iterator = Iterator\filter($emitter->getIterator(), function ($value) use (&$invoked) {
                 $invoked = true;
             });
 
-            $this->assertInstanceOf(Stream::class, $stream);
+            $this->assertInstanceOf(Iterator::class, $iterator);
 
-            $emitter->resolve();
+            $emitter->complete();
         });
 
         $this->assertFalse($invoked);
     }
 
     public function testValuesEmitted() {
-        $count = 0;
-        $values = [1, 2, 3];
-        $results = [];
-        $expected = [1, 3];
-        Loop::run(function () use (&$results, &$result, &$count, $values) {
+        Loop::run(function () {
+            $count = 0;
+            $values = [1, 2, 3];
+            $expected = [1, 3];
             $producer = new Producer(function (callable $emit) use ($values) {
                 foreach ($values as $value) {
                     yield $emit($value);
                 }
             });
 
-            $stream = Stream\filter($producer, function ($value) use (&$count) {
+            $iterator = Iterator\filter($producer, function ($value) use (&$count) {
                 ++$count;
                 return $value & 1;
             });
 
-            $stream->onEmit(function ($value) use (&$results) {
-                $results[] = $value;
-            });
-
-            $stream->onResolve(function ($exception, $value) use (&$result) {
-                $result = $value;
-            });
+            while (yield $iterator->advance()) {
+                $this->assertSame(\array_shift($expected), $iterator->getCurrent());
+            }
+            $this->assertSame(3, $count);
         });
-
-        $this->assertSame(\count($values), $count);
-        $this->assertSame($expected, $results);
     }
 
     /**
      * @depends testValuesEmitted
      */
     public function testCallbackThrows() {
-        $values = [1, 2, 3];
-        $exception = new \Exception;
-        Loop::run(function () use (&$reason, $values, $exception) {
+        Loop::run(function () {
+            $values = [1, 2, 3];
+            $exception = new \Exception;
             $producer = new Producer(function (callable $emit) use ($values) {
                 foreach ($values as $value) {
                     yield $emit($value);
                 }
             });
 
-            $stream = Stream\filter($producer, function () use ($exception) {
+            $iterator = Iterator\filter($producer, function () use ($exception) {
                 throw $exception;
             });
 
-            $stream->onEmit(function ($value) use (&$results) {
-                $results[] = $value;
-            });
-
-            $callback = function ($exception, $value) use (&$reason) {
-                $reason = $exception;
-            };
-
-            $stream->onResolve($callback);
+            try {
+                while (yield $iterator->advance()) {
+                    $iterator->getCurrent();
+                }
```

Comment:
Yep, there's several other places with the same pattern. A caveat is that `catch` will catch the test failure, so maybe I need a specific exception class too.